internal/rbd/rbd_util.go

```diff
@@ -843,62 +939,79 @@ func (rv *rbdVolume) getImageInfo() error {
+func (rv *rbdVolume) getCloneDepth(ctx context.Context) (int, error) {
+	depth := 0
+	imageName := rv.RbdImageName
+	poolName := rv.Pool
+	for {
+		if imageName == "" {
+			return depth, nil
+		}
+		vol := rbdVolume{
+			Pool:         poolName,
+			RbdImageName: imageName,
+		}
+		err := vol.getImageInfo()
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to check depth on image: %s"), imageName, err)
+			return depth, err
+		}
+		if vol.ParentName != "" {
+			depth++
+		}
+		imageName = vol.ParentName
+	}
 }
 
-/*
-getSnapInfo queries rbd about the snapshots of the given image and returns its metadata, and
-returns ErrImageNotFound if provided image is not found, and ErrSnapNotFound if provided snap
-is not found in the images snapshot list
-*/
-func (rs *rbdSnapshot) getSnapInfo(ctx context.Context, monitors string, cr *util.Credentials) (snapInfo, error) {
-	// rbd --format=json snap ls [image-spec]
-
+func (rv *rbdVolume) flattenRbdImage(ctx context.Context, cr *util.Credentials) error {
 	var (
-		snpInfo snapInfo
-		snaps   []snapInfo
+		depth int
+		err   error
 	)
 
-	stdout, stderr, err := util.ExecCommand(
-		"rbd",
-		"-m", monitors,
-		"--id", cr.ID,
-		"--keyfile="+cr.KeyFile,
-		"-c", util.CephConfigPath,
-		"--format="+"json",
-		"snap", "ls", rs.imageSpec())
+	depth, err = rv.getCloneDepth(ctx)
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed getting snap (%s) information from image (%s): (%s)"),
-			rs.RbdSnapName, rs.imageSpec(), err)
-		if strings.Contains(string(stderr), "rbd: error opening image "+rs.RbdImageName+
-			": (2) No such file or directory") {
-			return snpInfo, ErrImageNotFound{rs.imageSpec(), err}
-		}
-		return snpInfo, err
-	}
-
-	err = json.Unmarshal(stdout, &snaps)
-	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed to parse JSON output of image snap list (%s): (%s)"), rs.imageSpec(), err)
-		return snpInfo, fmt.Errorf("unmarshal failed: %+v. raw buffer response: %s", err, string(stdout))
+		return err
 	}
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rv)
 
-	for _, snap := range snaps {
-		if snap.Name == rs.RbdSnapName {
-			return snap, nil
+	if (depth != 0 && flattenOnCreate) || (depth >= rbdHardMaxCloneDepth) || (depth >= rbdSoftMaxCloneDepth) {
+		args := []string{"flatten", rv.Pool + "/" + rv.RbdImageName, "--id", cr.ID, "--keyfile=" + cr.KeyFile, "-m", rv.Monitors}
+		ns, err := addRbdManagerTask(ctx, rv, args)
+		if ns {
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to add task flatten for %s : %v"), rv, err)
+				return err
+			}
+			if flattenOnCreate || depth >= rbdHardMaxCloneDepth {
+				return ErrFlattenInProgress{err: fmt.Errorf("flatten is in progress for image %s", rv.RbdImageName)}
+			}
+		}
+		if !ns {
+			klog.Errorf(util.Log(ctx, "rbd task add flatten not supported will flatten image once hardlimit is reached: %v"), err)
```

Comment:
maybe quote the command, or describe it like "task manager does not support flatten, image will be .."